src/vanilla/store.ts

```diff
@@ -418,118 +426,156 @@ const buildStore = (
+          stack.push([d, atomMap.get(d)!])
         }
       }
     }
-    return [sorted, visited]
+    return sorted.reverse()
   }
 
-  const recomputeDependents = <Value>(
-    pending: Pending,
-    atom: Atom<Value>,
-    atomState: AtomState<Value>,
-  ) => {
-    // Step 1: traverse the dependency graph to build the topsorted atom list
-    // We don't bother to check for cycles, which simplifies the algorithm.
-    const [topsortedAtoms, markedAtoms] = getSortedDependents(
-      pending,
-      atom,
-      atomState,
-    )
-
-    // Step 2: use the topsorted atom list to recompute all affected atoms
-    // Track what's changed, so that we can short circuit when possible
-    const changedAtoms = new Set<AnyAtom>([atom])
-    for (let i = topsortedAtoms.length - 1; i >= 0; --i) {
-      const [a, aState, prevEpochNumber] = topsortedAtoms[i]!
-      let hasChangedDeps = false
-      for (const dep of aState.d.keys()) {
-        if (dep !== a && changedAtoms.has(dep)) {
-          hasChangedDeps = true
-          break
-        }
-      }
-      if (hasChangedDeps) {
-        readAtomState(pending, a, markedAtoms)
+  const recomputeDependents = (pending: Pending, rootAtoms: Set<AnyAtom>) => {
+    if (rootAtoms.size === 0) {
+      return
+    }
+    const hasChangedDeps = (aState: AtomState) =>
+      Array.from(aState.d.keys()).some((d) => rootAtoms.has(d))
+    // traverse the dependency graph to build the topsorted atom list
+    for (const a of getSortedDependents(pending, rootAtoms)) {
+      // use the topsorted atom list to recompute all affected atoms
+      // Track what's changed, so that we can short circuit when possible
+      const aState = getAtomState(a)
+      const prevEpochNumber = aState.n
+      if (isPendingRecompute(a) || hasChangedDeps(aState)) {
+        readAtomState(pending, a)
         mountDependencies(pending, a, aState)
         if (prevEpochNumber !== aState.n) {
-          addPendingAtom(pending, a, aState)
-          changedAtoms.add(a)
+          markRecomputePending(pending, a, aState)
         }
       }
-      markedAtoms.delete(a)
+      markRecomputeComplete(pending, a, aState)
     }
   }
 
+  const recomputeDependencies = (pending: Pending, a: AnyAtom) => {
+    if (!isPendingRecompute(a)) {
+      return
+    }
+    const getDependencies = (_: unknown, aState: AtomState) => aState.d.keys()
+    const dependencies = Array.from(getDeep(getDependencies, [a]).keys())
+    const dirtyDependencies = new Set(dependencies.filter(isPendingRecompute))
+    recomputeDependents(pending, dirtyDependencies)
+  }
+
   const writeAtomState = <Value, Args extends unknown[], Result>(
     pending: Pending,
     atom: WritableAtom<Value, Args, Result>,
     ...args: Args
   ): Result => {
     let isSync = true
-    const getter: Getter = <V>(a: Atom<V>) =>
-      returnAtomValue(readAtomState(pending, a))
+    const getter: Getter = <V>(a: Atom<V>) => {
+      recomputeDependencies(pending, atom)
```

Comment:
I'm not sure why this is necessary.

Reply:
Perhaps I should leave a comment here. Before we get any atom, we need to recompute the dependents of any of its dependencies that have been marked as dirty. Tonensure we are returning the latest value. This is the exceltikn to the lazy recomputeDependents rule.

Example:
```ts
const a = atom(0)
const b = atom(0)
const c = atom((get) => [get(a), get(b)])
const d = atom((get) => get(c))
const e = atom((get) => get(a))
const w = atom(null, (get, set) => {
  get(d) // [0, 0]

  // does not immediately recompute
  // marks a as dirty
  set(a, 1)

  // setA: dependencies of d are {c,a,b}
  // setB: dirty atoms of setA are {a}
  // setC: dependents setB are {a,c,e}
  // setD: intersection of setC and setB are {a,c}
  // recompute setD
  get(d) // [1, 0]

  // recompute remaining dependents of dirty atoms in flushPending
})
store.set(w)
```

Reply:
I'm not convinced but maybe I need to play with it to get better understanding. It's my problem.

Reply:
I made this graphic to help me reason about this better.
![image](https://github.com/user-attachments/assets/c03f0d73-1cd3-441d-91f5-0416f2b2c16b)